Mathlib/Topology/Connected/RealIntervals.lean

```diff
@@ -0,0 +1,560 @@
+
+lemma Ioc_Ioo_with_endpoint {a b : ℝ} (hab : a < b) :
+    Ioc a b = (Ioo a b) ∪ {b} := by
+  apply Subset.antisymm
+  · intro x ⟨a_lt_x, x_le_b⟩
+    by_cases xb : x = b
+    · right; exact xb
+    have : x < b := Ne.lt_of_le xb x_le_b
+    left; exact ⟨a_lt_x, this⟩
+  · rintro x (x_Ioo | x_b)
+    · exact ⟨x_Ioo.1, LT.lt.le x_Ioo.2⟩
+    · exact ⟨(by rw [x_b]; exact hab), Eq.le x_b⟩
+
+section X_section
+variable
+  {X : Set ℝ}
+
+lemma bdd_subset_Icc {infX supX : ℝ}
+    (h_infX : IsGLB X infX) (h_supX : IsLUB X supX) :
+    X ⊆ Icc infX supX := by
+  intro x xX
+  exact ⟨h_infX.1 xX, h_supX.1 xX⟩
+
+lemma x_lt_excluded_supX {x supX : ℝ} (xX: x ∈ X)
+    (h_supX: IsLUB X supX) (supX_X : supX ∉ X) : x < supX := by
+  have : x ≠ supX := by
+    intro x_eq_supX
+    rw [← x_eq_supX] at supX_X
+    contradiction
+  exact Ne.lt_of_le this (h_supX.1 xX)
+
+lemma excluded_infX_lt_x {x infX : ℝ} (xX: x ∈ X)
+    (h_infX: IsGLB X infX) (infX_X : infX ∉ X) : infX < x := by
+  have : infX ≠ x := by
+    intro x_eq_supX
+    rw [x_eq_supX] at infX_X
+    contradiction
+  exact Ne.lt_of_le this (h_infX.1 xX)
+
+section conn_section
+variable
+  (conn : IsConnected X)
+
+-- connected in ℝ means order-connected.
+lemma conn_has_Icc
+    (a : ℝ) (aX : a ∈ X) (b : ℝ) (bX : b ∈ X) : Icc a b ⊆ X := by
+  have : OrdConnected X :=
+    IsPreconnected.ordConnected (IsConnected.isPreconnected conn)
+  exact Set.Icc_subset X aX bX
+
+lemma connected_bddAbove_subset_contains_Ioo
+    {supX x: ℝ} (h_supX : IsLUB X supX) (xX : x ∈ X) :
+    Ioo x supX ⊆ X := by
+  intro y ⟨x_lt_y, y_lt_supX⟩
+  by_cases h: ∃ z ∈ X, y ≤ z
+  · rcases h with ⟨z, zX, y_le_z⟩
+    have : Icc x z ⊆ X := conn_has_Icc conn x xX z zX
+    exact this ⟨LT.lt.le x_lt_y, y_le_z⟩
+  · push_neg at h
+    have : y ∈ upperBounds X := fun z ↦ fun zX ↦ LT.lt.le (h z zX)
+    have : supX ≤ y := h_supX.2 this
+    linarith
+
+lemma connected_bddBelow_subset_contains_Ioo
+    {infX : ℝ} {x : ℝ} (h_infX : IsGLB X infX) (xX : x ∈ X) :
+    Ioo infX x ⊆ X := by
+  intro y ⟨infX_lt_y, y_lt_x⟩
+  by_cases h: ∃ z ∈ X, z ≤ y
+  · rcases h with ⟨z, zX, z_le_y⟩
+    have : Icc z x ⊆ X := conn_has_Icc conn z zX x xX
+    exact this ⟨z_le_y, LT.lt.le y_lt_x⟩
+  · push_neg at h
+    have : y ∈ lowerBounds X := fun z ↦ fun zX ↦ LT.lt.le (h z zX)
+    have : y ≤ infX := h_infX.2 this
+    linarith
+
+section inf_sup_section
+variable
+  {infX supX : ℝ}
+  (h_infX : IsGLB X infX)
+  (h_supX : IsLUB X supX)
+
+lemma connected_bdd_subset_contains_Ioo :
+    Ioo infX supX ⊆ X := by
+  let ⟨z, zX⟩ := IsConnected.nonempty conn
+  have h₁ : Ioo infX z ⊆ X := connected_bddBelow_subset_contains_Ioo conn h_infX zX
+  have h₂ : Ioo z supX ⊆ X := connected_bddAbove_subset_contains_Ioo conn h_supX zX
+  have h₃ : Ioo infX supX ⊆ Ioo infX z ∪ {z} ∪ Ioo z supX := by
+    intro x ⟨infX_lt_x, x_lt_supX⟩
+    rcases lt_trichotomy x z with (x_lt_z | x_eq_z | z_lt_z)
+    · left; left
+      exact ⟨infX_lt_x, x_lt_z⟩
+    · left; right
```

Comment:
I do wonder if this can be golfed (i.e., written more concisely); `left; left` sticks out to me. If it can, you're welcome to do it.